tools/ci-build/publisher/src/package.rs

```diff
@@ -192,18 +262,28 @@ pub async fn discover_manifests(path: PathBuf) -> Result<Vec<PathBuf>> {
 }
 
 /// Discovers and parses all Cargo.toml files that are packages (as opposed to being exclusively workspaces)
-pub async fn discover_packages(fs: Fs, path: PathBuf) -> Result<Vec<Package>> {
+pub async fn discover_packages<P: ParseIntoLenientVersion>(
+    fs: Fs,
+    path: PathBuf,
+) -> Result<Vec<Package>> {
     let manifest_paths = discover_manifests(path).await?;
-    read_packages(fs, manifest_paths).await
+    read_packages::<P>(fs, manifest_paths).await
 }
 
-/// Parses a semver version number and adds additional error context when parsing fails.
-pub fn parse_version(manifest_path: &Path, version: &str) -> Result<Version, Error> {
-    Version::parse(version)
+/// Parses `version` into `LenientVersion` and adds additional error context when parsing fails.
+pub fn parse_version<P: ParseIntoLenientVersion>(
+    manifest_path: &Path,
+    version: &str,
+) -> Result<LenientVersion, Error> {
+    P::ParsedT::from_str(version)
+        .map(|ver| ver.into())
         .map_err(|err| Error::InvalidCrateVersion(manifest_path.into(), version.into(), err.into()))
```

Comment:
it seems like this function should return the type you ask for? (Like `parse::<T>`)?

Reply:
The function returning what we ask for actually helped get rid of [try_into_semver](https://github.com/awslabs/smithy-rs/pull/3009/commits/000950d7a0996d94b40f5ca5c174bbf0bba30f4c#diff-2cd24b049e142d74c42052e9ba595284814576c4bcf07fe9427621fdfe4349b7R60-R65), so I like it. Addressed in 0a4ab75.